--- build-deps	2008-07-18 10:50:57.000000000 +0000
+++ build-deps	2008-07-18 10:49:10.000000000 +0000
@@ -968,6 +968,10 @@
 	$(LIBTOOL) --mode=compile $(CC) $(CFLAGS) $(EXTRA_CFLAGS) -c ./src/video/dummy/SDL_nullvideo.c  -o $@
 
 
+$(objects)/lcd-as-memframe.lo: ./src/video/ipod/lcd-as-memframe.S \
+
+	$(LIBTOOL) --mode=compile $(CC) $(CFLAGS) $(EXTRA_CFLAGS) -c ./src/video/ipod/lcd-as-memframe.S -o $@
+
 $(objects)/SDL_ipodvideo.lo: ./src/video/ipod/SDL_ipodvideo.c \
 	./include/SDL.h \
 	./include/SDL_active.h \
