--- Makefile	2008-07-18 10:50:54.000000000 +0000
+++ Makefile	2008-07-18 10:52:26.000000000 +0000
@@ -50,6 +50,9 @@
 LT_REVISION = 2
 LT_LDFLAGS  = -no-undefined -rpath $(DESTDIR)$(libdir) -release $(LT_RELEASE) -version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
+SOURCES+=./src/video/ipod/lcd-as-memframe.S
+OBJECTS+=$(objects)/lcd-as-memframe.lo
+
 all: $(srcdir)/configure Makefile $(objects) $(objects)/$(TARGET) $(objects)/$(SDLMAIN_TARGET)
 
 $(srcdir)/configure: $(srcdir)/configure.in
