Index: core/module.c
===================================================================
--- core/module.c	(revision 2399)
+++ core/module.c	(working copy)
@@ -135,13 +135,13 @@
     return 0;
 }
 
-#define MODULE_INF_FILE "Module"
+#define MODULE_INF_FILE "Module"
 static void load_modinf (PzModule *mod) 
 {
     char *buf = malloc (512);
 #ifdef IPOD
-    char *confdir = "/etc/podzilla";
-    char *moddir = "/etc/podzilla/modules";
+    char *confdir = "/sandbox";
+    char *moddir = "/sandbox/modules";
 #else
     char *confdir = malloc( MAXPATHLEN + 10 );
     char *moddir = malloc( MAXPATHLEN + 20 );
