Index: core/pz.h
===================================================================
--- core/pz.h	(revision 2423)
+++ core/pz.h	(working copy)
@@ -473,12 +473,15 @@
 /** File browser helper functions - browser.c **/
 TWindow *pz_browser_open (const char *path);
 TWidget *pz_browser_get_actions (const char *path); // returns a menu widget
+TWindow *new_textview_window(char *filename);
 void pz_browser_set_handler (int (*pred)(const char *), TWindow *(*handler)());
 void pz_browser_remove_handler (int (*pred)(const char *));
 void pz_browser_add_action (int (*pred)(const char *), ttk_menu_item *action); // action->data will be set to file's full name
 void pz_browser_remove_action (int (*pred)(const char *));
 void pz_exec(const char *file);
 void pz_execv(const char *path, char *const argv[]);
+void pz_exec_kill(const char *file);
+void pz_execv_kill(const char *path, char *const argv[]);
 
 
 /** Text input functions - input.c **/
