Index: modules/podwrite/podwrite.c
===================================================================
--- modules/podwrite/podwrite.c	(revision 2399)
+++ modules/podwrite/podwrite.c	(working copy)
@@ -86,7 +86,7 @@
 	PzWindow * ret;
 	TWidget * wid;
 	ret = pz_new_window(_("Save to..."), PZ_WINDOW_NORMAL);
-	wid = ti_new_standard_text_widget(10, 40, ret->w-20, 10+ttk_text_height(ttk_textfont), 0, (podwrite_filename?podwrite_filename:"/Notes/"), podwrite_save_callback);
+	wid = ti_new_standard_text_widget(10, 40, ret->w-20, 10+ttk_text_height(ttk_textfont), 0, (podwrite_filename?podwrite_filename:"/mnt/Notes/"), podwrite_save_callback);
 	ttk_add_widget(ret, wid);
 	ret = pz_finish_window(ret);
 	ti_widget_start(wid);
