Index: getLoader2Args.c
===================================================================
--- getLoader2Args.c	(revision 2397)
+++ getLoader2Args.c	(working copy)
@@ -46,6 +46,16 @@
 }
 */
 
+static char* getAutoExec (char* args) {
+  char *cmd;
+  cmd = strstr(args, "autoexec=");
+  if (cmd == NULL) {
+    return NULL;
+  } else {
+    return cmd + 9;
+  }
+}
+
 int main (int argc, char **argv) 
 {
 /* look for non-empty spaces:
@@ -65,7 +75,7 @@
 */
   char *args;
 
-  args = getArgs ((char*)0x80);
+  args = getAutoExec(getArgs ((char*)0x80));
   
   if (args) {
     puts (args);
Index: Makefile
===================================================================
--- Makefile	(revision 2397)
+++ Makefile	(working copy)
@@ -1,6 +1,6 @@
 getLoader2Args: getLoader2Args.o
-	arm-elf-gcc getLoader2Args.c -o getLoader2Args -elf2flt
+	arm-elf-gcc getLoader2Args.c -o getLoader2AutoExec -elf2flt
 #getLoader2Args.o: getLoader2Args.c
 clean:
-	rm -f getLoader2Args *.o
+	rm -f getLoader2AutoExec getLoader2AutoExec.gdb *.o
 .PHONY: clean
