diff -ur src/common/bigint.c src/common/bigint.c
--- src/common/bigint.c	2000-03-02 21:19:15.000000000 -0500
+++ src/common/bigint.c	2008-02-25 21:16:05.937500000 -0500
@@ -166,7 +166,7 @@
 
   if (ReadCfgItem(PI_CFG_FILE,"Files","SaveFile",
                   SaveFileName,Cfg_String,MAX_FILENAME)==0)
-     strcpy(SaveFileName,"piceb.sav");
+     strcpy(SaveFileName,"/opt/Tools/PiCalc/Output/piceb.sav");
 
   InitFFT(Len);
   InitFFTMul(Len);
@@ -219,7 +219,7 @@
   sprintf(Str,"Var%ld",VarsAllocated+1);
   if (ReadCfgItem(PI_CFG_FILE,"DiskNumbers",Str,
                   vp->FileName,Cfg_String,MAX_FILENAME)==0)
-     sprintf(vp->FileName,"pivar%ld.tmp",VarsAllocated+1);
+     sprintf(vp->FileName,"/opt/Tools/PiCalc/Output/pivar%ld.tmp",VarsAllocated+1);
 
   vp->fp=fopen(vp->FileName,"wb+");
   if (vp->fp==NULL)
@@ -1562,7 +1562,7 @@
 
     if (ReadCfgItem(PI_CFG_FILE,"Files","Pi_Outfile_Mask",
                     Mask,Cfg_String,MAX_FILENAME)==0)
-       strcpy(Mask,"pi%s.txt");
+       strcpy(Mask,"/opt/Tools/PiCalc/Output/pi%s.txt");
 
      sprintf(str,Mask,Num2Str(Len*RawIntDigits));
      OutPutFile=fopen(str,"wb");
@@ -1609,7 +1609,7 @@
     {char str[80];
     if (ReadCfgItem(PI_CFG_FILE,"Files","Pi_Outfile_Mask",
                     Mask,Cfg_String,MAX_FILENAME)==0)
-       strcpy(Mask,"pi%s.txt");
+       strcpy(Mask,"/opt/Tools/PiCalc/Output/pi%s.txt");
 
      sprintf(str,Mask,Num2Str(Len*RawIntDigits));
      OutPutFile=fopen(str,"w");
diff -ur src/common/cache.c src/common/cache.c
--- src/common/cache.c	2000-02-24 14:58:06.000000000 -0500
+++ src/common/cache.c	2008-02-25 21:15:38.406250000 -0500
@@ -245,14 +245,14 @@
 
 if (ReadCfgItem(PI_CFG_FILE,"Files","Convolution",FFTCash[0].Name,
                Cfg_String,MAX_FILENAME)==0)
-   strcpy(FFTCash[0].Name,"piconvl.tmp");
+   strcpy(FFTCash[0].Name,"/opt/Tools/PiCalc/Output/piconvl.tmp");
 
 for (x=1;x<MaxFFTCacheLines;x++)
   {
    sprintf(Str,"Cache%d",x);
    if (ReadCfgItem(PI_CFG_FILE,"FFT-Cache",Str,FFTCash[x].Name,
                   Cfg_String,MAX_FILENAME)==0)
-      sprintf(FFTCash[x].Name,"cache%d.tmp",x);
+      sprintf(FFTCash[x].Name,"/opt/Tools/PiCalc/Output/cache%d.tmp",x);
   }
 
 for (x=0;x<MaxFFTCacheLines;x++)
diff -ur src/common/ini.c src/common/ini.c
--- src/common/ini.c	2000-02-24 14:58:06.000000000 -0500
+++ src/common/ini.c	2008-02-25 21:15:20.046875000 -0500
@@ -156,7 +156,7 @@
  FILE *TempFile=NULL;
  int tries=0;
 
-strcpy(TempFileName,"tempfile.tmp");
+strcpy(TempFileName,"/opt/Tools/PiCalc/Output/tempfile.tmp");
 
 do {
     srand((unsigned int)time(NULL)+tries);
