Index: src/Makefile
===================================================================
--- src/Makefile	(revision 2420)
+++ src/Makefile	(working copy)
@@ -78,6 +78,10 @@
 endif
 endif
 
+ifdef SANSA
+MYCFLAGS += -DSANSA
+endif
+
 ifeq ($(GFXLIB),hotdog)
 MYCFLAGS += -DHDOG
 OBJS += hotdog.o
Index: src/ttk.c
===================================================================
--- src/ttk.c	(revision 2420)
+++ src/ttk.c	(working copy)
@@ -307,8 +307,13 @@
 	    ttk_screen->h = 110;
 	    ttk_screen->bpp = 2;
 	} else if (ver & TTK_POD_PHOTO) {
+#ifdef SANSA
+	    ttk_screen->w = 176;
+	    ttk_screen->h = 220;
+#else
 	    ttk_screen->w = 220;
 	    ttk_screen->h = 176;
+#endif
 	    ttk_screen->bpp = 16;
         } else if (ver & TTK_POD_NANO) {
             ttk_screen->w = 176;
@@ -887,7 +892,7 @@
     period = period ? MIN(MAX(period, 5), 250) : 20; // period: 40=2286Hz, 30=3024Hz, 20=4465Hz, 10=8547Hz
     duration = MIN(MAX(duration, 1), 1000); //duration in millieconds.
 	
-#ifdef IPOD
+#if defined(IPOD) && !defined(SANSA)
     if (ttk_get_podversion() & TTK_POD_PP502X) {
 	outl(inl(0x70000010) & ~0xc, 0x70000010);
 	outl(inl(0x6000600c) | 0x20000, 0x6000600c);    /* enable device */
@@ -924,6 +929,9 @@
 
 static long iPod_GetGeneration() 
 {
+#ifdef SANSA
+    return 0x6;
+#else
     int i;
     char cpuinfo[256];
     char *ptr;
@@ -940,6 +948,7 @@
     ptr = cpuinfo + i + 2;
     
     return strtol(ptr, NULL, 16);
+#endif
 }
 
 static int ttk_get_podversion_raw() 
@@ -1551,7 +1560,7 @@
 }
 
 
-#ifdef IPOD
+#if defined(IPOD) && !defined(SANSA)
 /*** The following LCD code is taken from Linux kernel uclinux-2.4.24-uc0-ipod2,
      file arch/armnommu/mach-ipod/fb.c. A few modifications have been made. ***/
 
Index: Makefile
===================================================================
--- Makefile	(revision 2420)
+++ Makefile	(working copy)
@@ -4,6 +4,9 @@
 
 all: build-dirs
 	@echo ">>> Building TTK..."
+ifdef SANSA
+	make -C build/ipod-sdl IPOD=1 SDL=1 SANSA=1 all
+else
 ifndef NOHDOG
 ifndef NOIPOD
 	make -C build/ipod-hdog IPOD=1 HDOG=1 all
@@ -28,6 +31,7 @@
 	make -C build/x11-mwin MWIN=1 all
 endif
 endif
+endif
 	@echo "<<< Done."
 	@echo
 
