Index: src/include/ttk.h
===================================================================
--- src/include/ttk.h	(revision 2423)
+++ src/include/ttk.h	(working copy)
@@ -29,6 +29,7 @@
 #define TTK_POD_4G        010
 #define TTK_POD_MINI_1G   020
 #define TTK_POD_PHOTO     040
+#define TTK_POD_SANSA     044
 #define TTK_POD_PP5020    070
 #define TTK_POD_MINI_2G   0100
 #define TTK_POD_MINI      0120
Index: src/ttk.c
===================================================================
--- src/ttk.c	(revision 2423)
+++ src/ttk.c	(working copy)
@@ -311,6 +311,10 @@
 	    ttk_screen->w = 220;
 	    ttk_screen->h = 176;
 	    ttk_screen->bpp = 16;
+	} else if (ver & TTK_POD_SANSA) {
+	    ttk_screen->w = 176;
+	    ttk_screen->h = 220;
+	    ttk_screen->bpp = 16;
         } else if (ver & TTK_POD_NANO) {
             ttk_screen->w = 176;
             ttk_screen->h = 132;
@@ -949,7 +953,7 @@
     static int ver;
     if (!ver) ver = iPod_GetGeneration();
     switch (ver >> 16) {
-    case 0x0: return 0;
+    case 0x0: return TTK_POD_SANSA;
     case 0x1: return TTK_POD_1G;
     case 0x2: return TTK_POD_2G;
     case 0x3: return TTK_POD_3G;
@@ -1785,6 +1789,9 @@
 	    ttk_quit();
 	    exit (1);
 	}
+	if (ver & TTK_POD_SANSA)
+		return; // SDL handles LCD code
+	
 	if (ver & (TTK_POD_1G|TTK_POD_2G|TTK_POD_3G)) {
 	    lcd_base = IPOD_OLD_LCD_BASE;
 	    lcd_rtc = IPOD_OLD_LCD_RTC;
