Index: core/Makefile
===================================================================
--- core/Makefile	(revision 2420)
+++ core/Makefile	(working copy)
@@ -10,6 +10,10 @@
 endif
 MYCFLAGS += -I/sw/include
 
+ifdef SANSA
+MYCFLAGS += -DSANSA
+endif
+
 OBJS= pz.o module.o ipod.o dialog.o header.o settings.o fonts.o \
       menu.o legacy.o gui.o browser.o vector.o oss.o appearance.o input.o \
       priority.o icons.o menuconf.o
