Index: core/pz.c
===================================================================
--- core/pz.c	(revision 2420)
+++ core/pz.c	(working copy)
@@ -319,7 +323,7 @@
 	struct stat statbuf;
 
 	/* find the last modified time of the settings file */
-	stat( "/etc/podzilla/podzilla.conf", &statbuf );
+	stat( "/opt/Tools/Encyclopodia/Conf/encyclopodia.conf", &statbuf );
 
 	/* convert timespec to timeval */
 	tv_s.tv_sec  = statbuf.st_mtime;
@@ -332,7 +336,7 @@
 void pz_touch_settings(void) 
 {
 #ifdef IPOD
-	close (open ("/etc/podzilla/podzilla.conf", O_WRONLY));
+	close (open ("/opt/Tools/Encyclopodia/Conf/encyclopodia.conf", O_WRONLY));
 #endif
 }
 
@@ -375,7 +379,7 @@
     unsigned long retaddr, off;
     const char *modfile = "Unknown";
     int i;
-    FILE *f = fopen ("podzilla.oops", "w");
+    FILE *f = fopen ("/opt/Tools/Encyclopodia/Misc/encyclopodia.oops", "w");
 
     asm ("mov %0, r11" : "=r" (FP) : );
 
@@ -558,7 +562,7 @@
 #endif
 
 #ifdef IPOD
-#define CONFIG_FILE "/etc/podzilla/podzilla.conf"
+#define CONFIG_FILE "/opt/Tools/Encyclopodia/Conf/encyclopodia.conf"
 #else
 #define CONFIG_FILE "config/podzilla.conf"
 #endif
@@ -621,13 +625,14 @@
 #define SET(x) pz_get_setting(pz_global_config,x)
 	if (!SET(WHEEL_DEBOUNCE))   pz_ipod_set (WHEEL_DEBOUNCE, 10);
 	if (!SET(CONTRAST))         pz_ipod_set (CONTRAST, initialContrast);
-	if (!SET(CLICKER))          pz_ipod_set (CLICKER, 1);
+	if (!SET(CLICKER))          pz_ipod_set (CLICKER, 0);
 	if (!SET(DSPFREQUENCY))     pz_ipod_set (DSPFREQUENCY, 0);
-	if (!SET(SLIDE_TRANSIT))    pz_ipod_set (SLIDE_TRANSIT, 1);
+	if (!SET(SLIDE_TRANSIT))    pz_ipod_set (SLIDE_TRANSIT, 2);
 	if (!SET(BACKLIGHT))        pz_ipod_set (BACKLIGHT, 1);
 	if (!SET(BACKLIGHT_TIMER))  pz_ipod_set (BACKLIGHT_TIMER, 3);
 	if (!SET(COLORSCHEME))
-		pz_set_string_setting (pz_global_config,COLORSCHEME, "mono.cs");
+		pz_set_string_setting (pz_global_config,COLORSCHEME, "skywards.cs");
+	if (!SET(GROUPED_MENUS))	pz_ipod_set (GROUPED_MENUS, 1);
 	pz_save_config (pz_global_config);
 	pz_ipod_fix_settings (pz_global_config);
 
@@ -640,7 +645,7 @@
 	/* load some fonts */
 	/* NOTE: we should probably do something if these fail! */
 	pz_load_font (&ttk_textfont, "Espy Sans", TEXT_FONT, pz_global_config);
-	pz_load_font (&ttk_menufont, "Chicago",   MENU_FONT, pz_global_config);
+	pz_load_font (&ttk_menufont, "GPalatino",   MENU_FONT, pz_global_config);
 
 	/* set up the menus and initialize the modules */
 	pz_menu_init();
@@ -648,10 +653,6 @@
 	pz_header_init();
 
 	/* sort the menus */
-	pz_menu_sort ("/Extras/Demos");
-	pz_menu_sort ("/Extras/Games");
-	pz_menu_sort ("/Extras/Utilities");
-	//pz_menu_sort ("/Extras/Applications");
 	if( pz_get_int_setting( pz_global_config, GROUPED_MENUS )) {
 		pz_menu_sort ("/Extras");
 		pz_menu_sort ("/");
