diff -ur common.h common.h
--- common.h	2008-06-27 14:26:31.953125000 -0400
+++ common.h	2008-06-27 14:26:47.140625000 -0400
@@ -106,6 +106,9 @@
 #ifndef IPOD_BUILD
   #include "SDL.h"
 #endif
+#ifdef ZEROSLACKR
+  #define ROOT "/opt/Emulators/igpSP"
+#endif
 
 #ifdef ARM_ARCH
   #define function_cc
diff -ur gui.c gui.c
--- gui.c	2008-06-27 14:26:32.000000000 -0400
+++ gui.c	2008-06-27 14:26:47.156250000 -0400
@@ -661,7 +661,11 @@
   u8 game_config_filename[512];
   u32 file_loaded = 0;
   u32 i;
+#ifdef ZEROSLACKR
+  change_ext(gamepak_filename, game_config_filename, ROOT "/Conf/", ".cfg");
+#else
   change_ext(gamepak_filename, game_config_filename, ".cfg");
+#endif
 
   file_open(game_config_file, game_config_filename, read);
 
@@ -725,7 +729,11 @@
   u8 config_path[512];
 
   #if (defined(PSP_BUILD) || defined(ARM_ARCH)) && !defined(_WIN32_WCE)
+  #ifdef ZEROSLACKR
+    sprintf(config_path, "%s%s", ROOT "/Conf/", GPSP_CONFIG_FILENAME);
+  #else
     sprintf(config_path, "%s/%s", main_path, GPSP_CONFIG_FILENAME);
+  #endif
   #else
     sprintf(config_path, "%s\\%s", main_path, GPSP_CONFIG_FILENAME);
   #endif
@@ -800,7 +808,11 @@
   u8 game_config_filename[512];
   u32 i;
 
+#ifdef ZEROSLACKR
+  change_ext(gamepak_filename, game_config_filename, ROOT "/Conf/", ".cfg");
+#else
   change_ext(gamepak_filename, game_config_filename, ".cfg");
+#endif
 
   file_open(game_config_file, game_config_filename, write);
 
@@ -833,7 +845,11 @@
   u8 config_path[512];
 
   #if (defined(PSP_BUILD) || defined(ARM_ARCH)) && !defined(_WIN32_WCE)
+  #ifdef ZEROSLACKR
+    sprintf(config_path, "%s%s", ROOT "/Conf/", GPSP_CONFIG_FILENAME);
+  #else
     sprintf(config_path, "%s/%s", main_path, GPSP_CONFIG_FILENAME);
+  #endif
   #else
     sprintf(config_path, "%s\\%s", main_path, GPSP_CONFIG_FILENAME);
   #endif
@@ -936,7 +952,11 @@
   u8 savestate_ext[16];
 
   sprintf(savestate_ext, "%d.svs", slot);
+#ifdef ZEROSLACKR
+  change_ext(gamepak_filename, name_buffer, ROOT "/Saves/", savestate_ext);
+#else
   change_ext(gamepak_filename, name_buffer, savestate_ext);
+#endif
 
   get_savestate_snapshot(name_buffer);
 }
@@ -946,7 +966,11 @@
   u8 savestate_ext[16];
 
   sprintf(savestate_ext, "%d.svs", slot);
+#ifdef ZEROSLACKR
+  change_ext(gamepak_filename, name_buffer, ROOT "/Saves/", savestate_ext);
+#else
   change_ext(gamepak_filename, name_buffer, savestate_ext);
+#endif
 }
 
 #ifdef PSP_BUILD
diff -ur ipod/Makefile ipod/Makefile
--- ipod/Makefile	2008-06-27 14:26:10.562500000 -0400
+++ ipod/Makefile	2008-06-27 14:26:47.156250000 -0400
@@ -36,6 +36,9 @@
 ifdef BENCHMARK
 CFLAGS     += -DBENCHMARK
 endif
+ifdef ZEROSLACKR
+CFLAGS     += -DZEROSLACKR
+endif
 CFLAGS     += -O3 -msoft-float -funsigned-char -fno-common                    \
               -fno-builtin -fomit-frame-pointer                               \
 			  -mapcs -march=armv4t -mcpu=arm7tdmi -elf2flt                    \
diff -ur ipod/ipod_conf.c ipod/ipod_conf.c
--- ipod/ipod_conf.c	2008-06-27 14:26:10.562500000 -0400
+++ ipod/ipod_conf.c	2008-06-27 14:27:36.265625000 -0400
@@ -120,6 +120,7 @@
 #endif
 
 static const char *igpsp_conf_filenames[] = {
+	"/opt/Emulators/igpSP/Conf/igpSP.ini",
 	"igpSP.ini",
 	"Conf/igpSP.ini",
 	"/igpSP.ini",
diff -ur main.c main.c
--- main.c	2008-06-27 14:26:32.015625000 -0400
+++ main.c	2008-06-27 14:26:47.156250000 -0400
@@ -184,6 +184,9 @@
 #ifdef ARM_ARCH
 void ChangeWorkingDirectory(char *exe)
 {
+#ifdef ZEROSLACKR
+  chdir(ROOT "/Roms");
+#else
 #ifndef _WIN32_WCE
   char *s = strrchr(exe, '/');
   if (s != NULL) {
@@ -192,6 +195,7 @@
     *s = '/';
   }
 #endif
+#endif
 }
 #endif
 
@@ -292,8 +296,12 @@
   sprintf(bios_filename, "%s/%s", main_path, "gba_bios.bin");
   if(load_bios(bios_filename) == -1)
 #else
+#ifdef ZEROSLACKR
+  if(load_bios(ROOT "/Data/gba_bios.bin") == -1)
+#else
   if(load_bios("gba_bios.bin") == -1)
 #endif
+#endif
   {
     gui_action_type gui_action = CURSOR_NONE;
 
@@ -1113,10 +1121,19 @@
 
 #endif
 
+#ifdef ZEROSLACKR
+void change_ext(u8 *src, u8 *buffer, u8 *prefix, u8 *extension)
+#else
 void change_ext(u8 *src, u8 *buffer, u8 *extension)
+#endif
 {
   u8 *dot_position;
+#ifdef ZEROSLACKR
+  strcpy(buffer, prefix);
+  strcat(buffer, src);
+#else
   strcpy(buffer, src);
+#endif
   dot_position = strrchr(buffer, '.');
 
   if(dot_position)
diff -ur main.h main.h
--- main.h	2008-06-27 14:26:32.015625000 -0400
+++ main.h	2008-06-27 14:26:47.156250000 -0400
@@ -240,7 +240,11 @@
 
 #endif
 
+#ifdef ZEROSLACKR
+void change_ext(u8 *src, u8 *buffer, u8 *prefix, u8 *extension);
+#else
 void change_ext(u8 *src, u8 *buffer, u8 *extension);
+#endif
 
 #endif
 
diff -ur memory.c memory.c
--- memory.c	2008-06-27 14:26:32.031250000 -0400
+++ memory.c	2008-06-27 14:26:47.171875000 -0400
@@ -2097,7 +2097,11 @@
   flash_device_id = FLASH_DEVICE_MACRONIX_64KB;
 
 #if (defined(PSP_BUILD) || defined(ARM_ARCH)) && !defined(_WIN32_WCE)
+#ifdef ZEROSLACKR
+  sprintf(config_path, "%s%s", ROOT "/Data/", CONFIG_FILENAME);
+#else
   sprintf(config_path, "%s/%s", main_path, CONFIG_FILENAME);
+#endif
 #else
   sprintf(config_path, "%s\\%s", main_path, CONFIG_FILENAME);
 #endif
@@ -2258,7 +2262,11 @@
 
     strcpy(backup_filename, name);
     strncpy(gamepak_filename, name, 512);
+#ifdef ZEROSLACKR
+    change_ext(gamepak_filename, backup_filename, ROOT "/Saves/", ".sav");
+#else
     change_ext(gamepak_filename, backup_filename, ".sav");
+#endif
 
     load_backup(backup_filename);
 
@@ -2272,7 +2280,11 @@
     load_game_config(gamepak_title, gamepak_code, gamepak_maker);
     load_game_config_file();
 
+#ifdef ZEROSLACKR
+    change_ext(gamepak_filename, cheats_filename, ROOT "/Cheats/", ".cht");
+#else
     change_ext(gamepak_filename, cheats_filename, ".cht");
+#endif
     add_cheats(cheats_filename);
 
     return 0;
