Index: src/Makefile
===================================================================
--- src/Makefile	(revision 2384)
+++ src/Makefile	(working copy)
@@ -81,12 +81,12 @@
 ifeq ($(GFXLIB),hotdog)
 MYCFLAGS += -DHDOG
 OBJS += hotdog.o
-MYCFLAGS += -I../../hotdog -I../../../hotdog `sdl-config --cflags`
+MYCFLAGS += -I../../hotdog -I../../../hotdog
 ifdef IPOD
 LIBSA += ../../../hotdog/ipod/libhotdog.a
 else
 LIBSA += ../../../hotdog/x11/libhotdog.a
-MYCFLAGS += -I/usr/include/SDL -I/usr/local/include/SDL `sdl-config --cflags`
+MYCFLAGS += -I/usr/include/SDL -I/usr/local/include/SDL
 LIBS += `sdl-config --libs` -lSDL_image -lpng -ljpeg
 endif
 LIBS += -lm
@@ -101,7 +101,7 @@
 LIBSA += ../libs/SDL/*.a
 LIBS += -lpthread -lm
 else
-MYCFLAGS += -I/usr/include/SDL -I/usr/local/include/SDL `sdl-config --cflags`
+MYCFLAGS += -I/usr/include/SDL -I/usr/local/include/SDL
 LIBS += `sdl-config --libs` -lSDL_image -lpng -ljpeg
 endif
 
Index: ttk.c
===================================================================
--- src/ttk.c	(revision 2384)
+++ src/ttk.c	(working copy)
@@ -25,7 +25,6 @@
 #include <string.h>
 #include <dirent.h>
 #include <sys/stat.h>
-#include <SDL.h>
 
 TWindowStack *ttk_windows = 0;
 ttk_font ttk_menufont, ttk_textfont;
