Index: hotdog_lcd.c
===================================================================
--- hotdog_lcd.c	(revision 2423)
+++ hotdog_lcd.c	(working copy)
@@ -272,6 +272,15 @@
 	}
 }
 
+#define LCD_FB_BASE_REG (*(volatile unsigned long *)(0xc2000028))
+extern void lcd_copy_buffer_rect(uint16 *dst, const uint16 *src, int width, int height);
+static void lcd_update_sansa_display(uint16 *fb, int sx, int sy, int width, int height)
+{
+	const uint16 *src = fb;
+	uint16 *dst = (uint16 *)(LCD_FB_BASE_REG & 0x0fffffff);
+	lcd_copy_buffer_rect(dst, src, 176 * 220, 1);
+}
+
 static long iPod_GetGeneration() 
 {
 	static long gen = 0;
@@ -308,6 +317,14 @@
 {
 	hw_ver = iPod_GetGeneration() >> 16;
 	switch (hw_ver) {
+	case 0x0: // Sansa e200
+		lcd_width = 176;
+		lcd_height = 220;
+		lcd_type = 4;
+		// lcd_base and rtc not used for now, but here anyway
+		//ipod_lcd_base = 0xc0001000;
+		//ipod_rtc = 0xcf001110;
+		break;
 	case 0xB: // video
 		lcd_width = 320;
 		lcd_height = 240;
@@ -390,6 +407,9 @@
 	case 3: // new mono (mini2g)
 		lcd_update_mono_display (fb, x, y, w, h);
 		break;
+	case 4: // Sansa e200
+		lcd_update_sansa_display (fb, x, y, w, h);
+		break;
 	case 5: // video
 		_HD_ARM_Update5G (fb, x, y, w, h);
 		break;
Index: Makefile
===================================================================
--- Makefile	(revision 2423)
+++ Makefile	(working copy)
@@ -66,7 +66,7 @@
           $(OBJDIR)/hotdog_canvas.o $(OBJDIR)/hotdog_surface.o \
           $(OBJDIR)/hotdog_ops.o $(OBJDIR)/hotdog_primitive.o $(OBJDIR)/zsort.o
 ifdef IPOD
-LIBOBJS += $(OBJDIR)/hotdog_lcd.o
+LIBOBJS += $(OBJDIR)/hotdog_lcd.o $(OBJDIR)/hotdog_lcd_sansa.o
 endif
 
 all: $(LIB)
