--- linux-2.4.32-ipod2/arch/armnommu/mach-ipod/audio.c	2008-03-26 19:29:42.343750000 -0400
+++ linux-2.4.32-ipod2/arch/armnommu/mach-ipod/audio.c	2008-03-26 20:45:36.390625000 -0400
@@ -1336,7 +1336,10 @@
 			return put_user(ipod_active_rec, (int *)arg);
 
 		case SOUND_MIXER_PCM:		/* codec output level */
-			val = (ipod_pcm_level - 0x2f) * 100 / 80;
+			if (codec_chip >= WM8758)
+				val = (ipod_pcm_level * 100) / 63;
+			else
+				val = (ipod_pcm_level - 0x2f) * 100 / 80;
 			val = val << 8 | val;
 			return put_user(val, (int *)arg);
 
