Index: linux-2.4.32-ipod2/Makefile
===================================================================
--- linux-2.4.32-ipod2/Makefile
+++ linux-2.4.32-ipod2/Makefile
@@ -28,7 +28,7 @@ HOSTCC  	= unset GCC_EXEC_PREFIX; gcc
 HOSTCFLAGS	= -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer
 
 # CROSS_COMPILE 	= m68k-elf-
-CROSS_COMPILE 	= arm-elf-
+CROSS_COMPILE 	= arm-uclinux-elf-
 # CROSS_COMPILE 	= h8300-elf-
 # CROSS_COMPILE 	= nios-elf-
 # CROSS_COMPILE		= e1-coff-
Index: linux-2.4.32-ipod2/arch/armnommu/Makefile
===================================================================
--- linux-2.4.32-ipod2/arch/armnommu/Makefile
+++ linux-2.4.32-ipod2/arch/armnommu/Makefile
@@ -57,8 +57,8 @@ LINKFLAGS	+= -EB
 LDFLAGS		+= -EB
 endif
 
-CFLAGS		+= $(apcs-y) $(arch-y) $(tune-y) -mshort-load-bytes -msoft-float
-AFLAGS		+= $(apcs-y) $(proc-y) -msoft-float -mno-fpu
+CFLAGS		+= $(apcs-y) $(arch-y) $(tune-y) -malignment-traps -msoft-float
+AFLAGS		+= $(apcs-y) $(proc-y) -msoft-float
 
 LIBGCC		:= $(shell $(CC) $(CFLAGS) --print-libgcc-file-name)
 
Index: linux-2.4.32-ipod0/include/asm-armnommu/arch-ipod/hardware.h
===================================================================
--- linux-2.4.32-ipod2/include/asm-armnommu/arch-ipod/hardware.h
+++ linux-2.4.32-ipod2/include/asm-armnommu/arch-ipod/hardware.h
@@ -100,7 +100,7 @@ struct sysinfo_t {
 
 extern unsigned ipod_get_hw_version(void);
 extern struct sysinfo_t *ipod_get_sysinfo(void);
-extern int ipod_is_pp5022();
+extern int ipod_is_pp5022(void);
 
 extern void ipod_i2c_init(void);
 extern int ipod_i2c_send_bytes(unsigned int addr, unsigned int len, unsigned char *data);
